ethcore/src/private_transactions/mod.rs

```diff
@@ -0,0 +1,819 @@
+
+/// Private transaction message call to the contract
+#[derive(Default, Debug, Clone, PartialEq, Eq)]
+pub struct PrivateTransaction {
+	/// Encrypted data
+	encrypted: Bytes,
+	/// Address of the contract
+	contract: Address,
+}
+
+impl Decodable for PrivateTransaction {
+	fn decode(d: &UntrustedRlp) -> Result<Self, DecoderError> {
+		if d.item_count()? != 2 {
+			return Err(DecoderError::RlpIncorrectListLen);
+		}
+		Ok(PrivateTransaction {
+			encrypted: d.val_at(0)?,
+			contract: d.val_at(1)?,
+		})
+	}
+}
+
+impl Encodable for PrivateTransaction {
+	fn rlp_append(&self, s: &mut RlpStream) {
+		s.begin_list(2);
+		s.append(&self.encrypted);
+		s.append(&self.contract);
+	}
+}
+
+fn initialization_vector() -> [u8; INIT_VEC_LEN] {
+	let mut result = [0u8; INIT_VEC_LEN];
+	let mut rng = OsRng::new().unwrap();
+	rng.fill_bytes(&mut result);
+	result
+}
+
+impl PrivateTransaction {
+	/// Compute hash on private transaction
+	pub fn hash(&self) -> H256 {
+		keccak(&*self.rlp_bytes())
+	}
+}
+
+/// Message about private transaction's signing
+#[derive(Default, Debug, Clone, PartialEq, Eq)]
+pub struct SignedPrivateTransaction {
+	/// Hash of the corresponding private transaction
+	private_transaction_hash: H256,
+	/// Signature of the validator
+	/// The V field of the signature
+	v: u64,
+	/// The R field of the signature
+	r: U256,
+	/// The S field of the signature
+	s: U256,
+}
+
+impl Decodable for SignedPrivateTransaction {
+	fn decode(d: &UntrustedRlp) -> Result<Self, DecoderError> {
+		if d.item_count()? != 4 {
+			return Err(DecoderError::RlpIncorrectListLen);
+		}
+		Ok(SignedPrivateTransaction {
+			private_transaction_hash: d.val_at(0)?,
+			v: d.val_at(1)?,
+			r: d.val_at(2)?,
+			s: d.val_at(3)?,
+		})
+	}
+}
+
+impl Encodable for SignedPrivateTransaction {
+	fn rlp_append(&self, s: &mut RlpStream) {
+		s.begin_list(4);
+		s.append(&self.private_transaction_hash);
+		s.append(&self.v);
+		s.append(&self.r);
+		s.append(&self.s);
+	}
+}
+
+impl SignedPrivateTransaction {
+	/// Construct a signed private transaction message
+	pub fn new(private_transaction_hash: H256, sig: Signature, chain_id: Option<u64>) -> Self {
+		SignedPrivateTransaction {
+			private_transaction_hash: private_transaction_hash,
+			r: sig.r().into(),
+			s: sig.s().into(),
+			v: sig.v() as u64 + if let Some(n) = chain_id { 35 + n * 2 } else { 27 },
+		}
+	}
+
+	/// 0 if `v` would have been 27 under "Electrum" notation, 1 if 28 or 4 if invalid.
+	pub fn standard_v(&self) -> u8 { match self.v { v if v == 27 || v == 28 || v > 36 => ((v - 1) % 2) as u8, _ => 4 } }
```

Comment:
I would prefer to have some explicit branches for 27 and 28. The `(v-1)%2` formula is confusing enough by itself...